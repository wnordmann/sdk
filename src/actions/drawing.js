/*
 * Copyright 2015-present Boundless Spatial Inc., http://boundlessgeo.com
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and limitations
 * under the License.
 */

/** @module actions/drawing
 * @desc Actions for interacting with the map.
 */

import {DRAWING} from '../action-types';
import {INTERACTIONS} from '../constants';

/** Action to start an interaction on the map.
 *  @param {string} sourceName The name of the source on which the action takes place.
 *  @param {string} drawingType The type of drawing interaction.
 *
 *  @returns {Object} An action object to pass to the reducer.
 */
export function startDrawing(sourceName, drawingType) {
  return {
    type: DRAWING.START,
    interaction: drawingType,
    sourceName,
  };
}

/** Short-hand action to start modify-feature
 *  @param {string} sourceName The name of the source to modify.
 *
 *  @returns {Object} Call to startDrawing()
 */
export function startModify(sourceName) {
  return startDrawing(sourceName, INTERACTIONS.modify);
}

/** Short-hand action to start select-feature
 *  @param {string} sourceName The name of the source on which the feature to select is found.
 *
 *  @returns {Object} Call to startDrawing()
 */
export function startSelect(sourceName) {
  return startDrawing(sourceName, INTERACTIONS.select);
}

/** Stop drawing / select / modify
 *  @returns {Object} An action object to pass to the reducer.
 */
export function endDrawing() {
  return {
    type: DRAWING.END,
  };
}

/** These are just aliases to end drawing.
 *  @returns {Object} Call to endDrawing().
 */
export function endModify() {
  return endDrawing();
}

/** These are just aliases to end drawing.
 *  @returns {Object} Call to endDrawing().
 */
export function endSelect() {
  return endDrawing();
}


/** Start measuring.
 *  @param {string} interaction Type of interation.
 *
 *  @returns {Object} Call to startDrawing().
 */
export function startMeasure(interaction) {
  return startDrawing(null, interaction);
}

/** Set a measurement feature.
 *
 *  This is called each time the feature is updated.
 *
 *  @param {Object} feature  The feature in WGS84.
 *  @param {number[]} segments Array of the incremental measurements in meters.
 *                    [] for a Point, [total_area] for a polygon.
 *
 *  @returns {Object} A measurement action.
 */
export function setMeasureFeature(feature, segments) {
  return {
    type: DRAWING.SET_MEASURE_FEATURE,
    feature,
    segments,
  };
}

/** Finalize a measurement feature.
 *
 *  This is called when the measure feature is done.
 *
 *  @returns {Object} A measurement action.
 */
export function finalizeMeasureFeature() {
  return {
    type: DRAWING.FINALIZE_MEASURE_FEATURE,
  };
}

/** Clear the measurement feature.
 *  @returns {Object} An action object to pass to the reducer.
 */
export function clearMeasureFeature() {
  return {
    type: DRAWING.CLEAR_MEASURE_FEATURE,
  };
}
/** Set the editing style.
 *  @param {Object} mbStyle The mapbox style to be used for the edit feature mode.
 *  @returns {Object} An action object to pass to the reducer.
 */
export function setEditStyle(mbStyle) {
  return {
    type: DRAWING.SET_EDIT_STYLE,
    editStyle: mbStyle
  };
}
/** Set the select style.
 *  @param {Object} mbStyle The mapbox style to be used for the select feature mode.
 *  @returns {Object} An action object to pass to the reducer.
 */
export function setSelectStyle(mbStyle) {
  return {
    type: DRAWING.SET_SELECT_STYLE,
    selectStyle: mbStyle
  };
}
/** Set the modify style.
 *  @param {Object} mbStyle The mapbox style to be used for the modify feature mode.
 *  @returns {Object} An action object to pass to the reducer.
 */
export function setModifyStyle(mbStyle) {
  return {
    type: DRAWING.SET_MODIFY_STYLE,
    modifyStyle: mbStyle
  };
}
