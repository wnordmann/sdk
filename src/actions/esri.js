/*
 * Copyright 2015-present Boundless Spatial Inc., http://boundlessgeo.com
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and limitations
 * under the License.
 */

/** @module actions/esri
 *  @desc Actions for dealing with Esri ArcGIS Rest Feature Services.
 */

import {ESRI} from '../action-types';

/** Esri add source action.
 *  @param {string} sourceName The source name to be added.
 *  @param {Object} options source definition options.
 *
 *  @returns {Object} Action object to pass to reducer.
 */
export function addSource(sourceName, options) {
  return {
    type: ESRI.ADD_SOURCE,
    sourceName,
    sourceDef: options,
  };
}

/** Esri remove source action.
 *  @param {string} sourceName The source name that will be removed.
 *
 *  @returns {Object} Action object to pass to reducer.
 */
export function removeSource(sourceName) {
  return {
    type: ESRI.REMOVE_SOURCE,
    sourceName,
  };
}
