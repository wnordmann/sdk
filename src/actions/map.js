/*
 * Copyright 2015-present Boundless Spatial Inc., http://boundlessgeo.com
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and limitations
 * under the License.
 */

import fetch from 'isomorphic-fetch';

/** @module actions/map
 * @desc Action Defintions for the map.
 */

import {MAP} from '../action-types';
import {TITLE_KEY, TIME_KEY} from '../constants';
import {encodeQueryObject} from '../util';

const sourceTypes = [
  'vector',
  'raster',
  'geojson',
  'image',
  'video',
  'canvas'
];

/** Action to update the center and zoom values in map state.
 *  @param {number[]} center Center coordinates.
 *  @param {number} zoom Zoom value.
 *
 *  @returns {Object} Action object to pass to reducer.
 */
export function setView(center, zoom) {
  return {
    type: MAP.SET_VIEW,
    view: {center, zoom},
  };
}

/** Action to zoom in on the map.
 *
 *  @returns {Object} Action object to pass to reducer.
 */
export function zoomIn() {
  return {
    type: MAP.ZOOM_IN,
  };
}

/** Action to zoom out on the map.
 *
 *  @returns {Object} Action object to pass to reducer.
 */
export function zoomOut() {
  return {
    type: MAP.ZOOM_OUT,
  };
}

/** Action to set the zoom level.
 *  @param {number} zoom Zoom value.
 *
 *  @returns {Object} Action object to pass to reducer.
 */
export function setZoom(zoom) {
  return {
    type: MAP.SET_ZOOM,
    zoom,
  };
}

/** Action to update the map name in map state.
 *  @param {string} name Map name.
 *
 *  @returns {Object} Action object to pass to reducer.
 */
export function setMapName(name) {
  return {
    type: MAP.SET_NAME,
    name,
  };
}

/** Action to update the map bearing value in map state.
 *  @param {number} degrees Bearing value to set in degrees.
 *
 *  @returns {Object} Action object to pass to reducer.
 */
export function setRotation(degrees) {
  return {
    type: MAP.SET_ROTATION,
    degrees,
  };
}

/** Action to add a layer object in the map state.
 *  @param {Object} layerDef Layer properties.
 *  @param {string} layerTitle Title of the layer to be added.
 *  @param {string} positionId String id for the layer.
 *
 *  @returns {Object} Action object to pass to reducer.
 */
export function addLayer(layerDef, layerTitle, positionId) {
  return {
    type: MAP.ADD_LAYER,
    layerDef,
    layerTitle,
    positionId,
  };
}

/** Action to add a source object in the map state.
 *  @param {string} sourceName Name of the source to be added.
 *  @param {Object} sourceDef Source properties.
 *
 *  @returns {Object} Action object to pass to reducer.
 */
export function addSource(sourceName, sourceDef) {
  if (sourceTypes.indexOf(sourceDef.type) === -1) {
    throw (new Error('Invalid source type: ' + sourceDef.type + '.  Valid source types are ' + sourceTypes.toString()));
  }
  return {
    type: MAP.ADD_SOURCE,
    sourceName,
    sourceDef,
  };
}

/** Action to remove a layer object in the map state.
 *  @param {string} layerId String id for the layer.
 *
 *  @returns {Object} Action object to pass to reducer.
 */
export function removeLayer(layerId) {
  return {
    type: MAP.REMOVE_LAYER,
    layerId,
  };
}

/** Action to remove a source object in the map state.
 *  @param {string} sourceName Name of the source to be added.
 *
 *  @returns {Object} Action object to pass to reducer.
 */
export function removeSource(sourceName) {
  return {
    type: MAP.REMOVE_SOURCE,
    sourceName,
  };
}

/** Action to update a layer object in the map state.
 *  @param {string} layerId String id for the layer to be updated.
 *  @param {Object} layerDef Layer properties.
 *
 *  @returns {Object} Action object to pass to reducer.
 */
export function updateLayer(layerId, layerDef) {
  return {
    type: MAP.UPDATE_LAYER,
    layerId,
    layerDef,
  };
}

/** Action to update cluster status in the map state.
 *  @param {string} sourceName Name of the source to be added.
 *  @param {boolean} isClustered Is the source clustered?
 *
 *  @returns {Object} Action object to pass to reducer.
 */
export function clusterPoints(sourceName, isClustered) {
  return {
    type: MAP.CLUSTER_POINTS,
    sourceName,
    cluster: isClustered,
  };
}

/** Set the radius of a clustering layer.
 *
 *  When set to a layer without clustering this will
 *  have no effect.
 *
 *  @param {Object} sourceName Name of the source on which the features are clustered.
 *  @param {number} radius Cluster radius.
 *
 *  @returns {Object} Action object to pass to reducer.
 */
export function setClusterRadius(sourceName, radius) {
  return {
    type: MAP.SET_CLUSTER_RADIUS,
    sourceName,
    radius,
  };
}

/** Add features to a source on the map.
 *
 *  @param {Object} sourceName Name of the source on which the features will be added.
 *  @param {Object[]} features An array of features to add.
 *  @param {number} position The position at which to add the features.
 *
 *  @returns {Object} Action object to pass to reducer.
 */
export function addFeatures(sourceName, features, position = -1) {
  return {
    type: MAP.ADD_FEATURES,
    sourceName,
    features,
    position,
  };
}

/** Remove features from a source on the map.
 *
 *  @param {Object} sourceName Name of the source on which the features will be removed.
 *  @param {string[]} filter Rule determining which features will be removed.
 *  See https://www.mapbox.com/mapbox-gl-js/style-spec/#types-filter.
 *
 *  @returns {Object} Action object to pass to reducer.
 */
export function removeFeatures(sourceName, filter) {
  return {
    type: MAP.REMOVE_FEATURES,
    sourceName,
    filter,
  };
}

/** Change the visibility of a given layer in the map state.
 *
 *  @param {string} layerId String id for the layer.
 *  @param {boolean} visibility Should the layer be visible?
 *
 *  @returns {Object} Action object to pass to reducer.
 */
export function setLayerVisibility(layerId, visibility) {
  return {
    type: MAP.SET_LAYER_VISIBILITY,
    layerId,
    visibility,
  };
}

/** Set a layer visible in a mutually exclusive group.
 *
 *  @param {string} layerId String id for the layer to turn on.
 *  @param {string} groupId String id for the layer group.
 *
 *  @returns {Object} Action object to pass to reducer.
 */
export function setLayerInGroupVisible(layerId, groupId) {
  return {
    type: MAP.SET_LAYER_IN_GROUP_VISIBLE,
    layerId,
    groupId,
  };
}

/** Change the metadata object in a given layer in the map state.
 *
 *  @param {string} layerId String id for the layer.
 *  @param {string} itemName A key.
 *  @param {number} itemValue A value.
 *
 *  @returns {Object} Action object to pass to reducer.
 */
export function setLayerMetadata(layerId, itemName, itemValue) {
  return {
    type: MAP.SET_LAYER_METADATA,
    layerId,
    key: itemName,
    value: itemValue,
  };
}

/** Set a layer's title in the map state.
 *
 *  @param {string} layerId String id for the layer.
 *  @param {string} title A string representing a layer title.
 *
 *  @returns {Object} Call to setLayerMetadata for a title key.
 */
export function setLayerTitle(layerId, title) {
  return setLayerMetadata(layerId, TITLE_KEY, title);
}

/** Set a layer's time property in the map state.
 *
 *  @param {string} layerId String id for the layer.
 *  @param {*} time The value of time assigned to the layer.
 *
 *  @returns {Object} Call to setLayerMetadata for a time key.
 */
export function setLayerTime(layerId, time) {
  return setLayerMetadata(layerId, TIME_KEY, time);
}

/** Receive a Mapbox GL style object.
 *
 *  @param {Object} context Mapbox GL style object to populate the map state.
 *
 *  @returns {Object} Action object to pass to reducer.
 */
export function receiveContext(context) {
  return {
    type: MAP.RECEIVE_CONTEXT,
    context,
  };
}

/** Thunk action creator to set the map state from a provided context object.
 *
 *  @param {Object} options A context object that must provide a Mapbox GL json
 *  object either via a json property or a from a url fetch.
 *
 *  @returns {Promise} A Promise object.
 */
export function setContext(options) {
  return (dispatch) => {
    if (options.url) {
      return fetch(options.url)
        .then(
          response => response.json(),
          error => console.error('An error occured.', error),
        )
        .then(json =>
          dispatch(receiveContext(json)),
        );
    } else if (options.json) {
      return new Promise((resolve) => {
        dispatch(receiveContext(options.json));
        resolve();
      });
    }
    return new Promise((resolve, reject) => {
      reject('Invalid option for setContext. Specify either json or url.');
    });
  };
}

/** Rearrange a layer in the list.
 *
 *  @param {string} layerId the layer to move.
 *  @param {string} targetLayerId The ID of the layer to move the layerId BEFORE.
 *                       When null or undefined, moves the layer to the end
 *                       of the list.
 *
 *  @returns {Object} An action object.
 */
export function orderLayer(layerId, targetLayerId) {
  return {
    type: MAP.ORDER_LAYER,
    layerId,
    targetId: targetLayerId,
  };
}

/** Set the sprite for the map.
 *
 *  @param {string} spriteRoot The URI to the sprite data without the .json/.png suffix.
 *
 *  @returns {Object} An action object.
 */
export function setSprite(spriteRoot) {
  return {
    type: MAP.SET_SPRITE,
    sprite: spriteRoot,
  };
}

/** Update the map's metadata.
 *
 *  @param {Object} metadata An object containing new/updated metadata.
 *
 *  @returns {Object} An action object.
 */
export function updateMetadata(metadata) {
  return {
    type: MAP.UPDATE_METADATA,
    metadata,
  };
}

/** Manually update a source.
 *
 *  @param {string} sourceName The name of the source to be updated.
 *  @param {Object} update The changes to the sourceDef to apply.
 *
 *  @returns {Object} An action object to pass to the reducer
 */
export function updateSource(sourceName, update) {
  return {
    type: MAP.UPDATE_SOURCE,
    sourceName,
    sourceDef: update,
  };
}

/** Set the time of the map.
 *
 *  @param {string} time An ISO date time string.
 *
 *  @returns {Object} An action object.
 */
export function setMapTime(time) {
  let metadata = {};
  metadata[TIME_KEY] = time;
  return updateMetadata(metadata);
}

/** Create an action for moving a group
 *
 *  @param {string} group Group name.
 *  @param {string} layerId Layer id of the new place for the layer.
 *
 *  @returns {object} An action object.
 */
export function moveGroup(group, layerId) {
  return {
    type: MAP.MOVE_GROUP,
    placeAt: layerId,
    group,
  };
}

/** Add a WMS source.
 *
 * @param {string} sourceId - new ID for the source.
 * @param {string} serverUrl - URL for the service.
 * @param {string} layerName - WFS feature type.
 * @param {Object} options - Optional settings. Honors: accessToken, projection, asVector, tileSize
 *
 * @returns {Object} Action to create a new source.
 */
export function addWmsSource(sourceId, serverUrl, layerName, options = {}) {
  const tile_size = options.tileSize ? options.tileSize : 256;
  const projection = options.projection ? options.projection : 'EPSG:3857';
  // default behaviour is vector
  const format = (options.asVector !== false) ? 'application/x-protobuf;type=mapbox-vector' : 'image/png';

  const params = {
    'SERVICE': 'WMS',
    'VERSION': '1.3.0',
    'REQUEST': 'GetMap',
    'FORMAT': format,
    'TRANSPARENT': 'TRUE',
    'LAYERS': layerName,
    'WIDTH': tile_size,
    'HEIGHT': tile_size,
    'CRS': projection,
  };

  if (options.accessToken) {
    params['ACCESS_TOKEN'] = options.accessToken;
  }

  // the BBOX is not escaped because the "{" and "}" are used for string
  // substitution in the library.
  const url_template = `${serverUrl}?${encodeQueryObject(params)}&BBOX={bbox-epsg-3857}`;

  if (options.asVector !== false) {
    return addSource(sourceId, {
      type: 'vector',
      url: url_template,
    });
  } else {
    return addSource(sourceId, {
      type: 'raster',
      tileSize: tile_size,
      tiles: [
        url_template,
      ],
    });
  }
}

/** Add a WFS / GeoJSON source.
 *
 * @param {string} sourceId - new ID for the source.
 * @param {string} serverUrl - URL for the service.
 * @param {string} featureType - WFS feature type.
 * @param {Object} options - Optional settings. Honors: accessToken
 *
 * @returns {Object} Action to create a new source.
 */
export function addWfsSource(sourceId, serverUrl, featureType, options = {}) {
  const params = {
    'SERVICE': 'WFS',
    'VERSION': '1.1.0',
    // projection is always fixed for WFS sources as they
    // are reprojected on the client.
    'SRSNAME': 'EPSG:4326',
    'REQUEST': 'GetFeature',
    'TYPENAME': featureType,
    'OUTPUTFORMAT': 'JSON',
  };

  if (options.accessToken) {
    params['ACCESS_TOKEN'] = options.accessToken;
  }

  return addSource(sourceId, {
    type: 'geojson',
    data: `${serverUrl}?${encodeQueryObject(params)}`
  });
}

/** Add a Mapbox Vector Tile layer from a TMS service.
 *
 * @param {string} sourceId - new ID for the source.
 * @param {string} serverUrl - URL for the service.
 * @param {string} layerName - Name of the layer.
 * @param {Object} options - Optional settings. Honors: accessToken
 *
 * @returns {Object} Action to create a new source.
 */
export function addTmsSource(sourceId, serverUrl, layerName, options = {}) {
  const token_string = options.accessToken ? `access_token=${options.accessToken}` : '';

  const url = `${serverUrl}/gwc/service/tms/1.0.0/${layerName}@EPSG%3A3857@pbf/{z}/{x}/{-y}.pbf?${token_string}`;

  return addSource(sourceId, {
    type: 'vector',
    url,
  });
}
