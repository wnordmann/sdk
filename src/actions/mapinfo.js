/*
 * Copyright 2015-present Boundless Spatial Inc., http://boundlessgeo.com
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and limitations
 * under the License.
 */

/** @module actions/mapinfo
 *  @desc Actions for setting info about the map.
 */

import {MAPINFO} from '../action-types';

/** Action to set the map size.
 *  @param {number[]} size Map size in pixels.
 *
 *  @returns {Object} Action object to pass to reducer.
 */
export function setMapSize(size) {
  return {
    type: MAPINFO.SET_SIZE,
    size,
  };
}

/** Action to set the current mouse position.
 *  @param {Object} lngLat The longitude latitude object.
 *  @param {number[]} coordinate Coordinate pair in map projection.
 *
 *  @returns {Object} Action object to pass to reducer.
 */
export function setMousePosition(lngLat, coordinate) {
  return {
    type: MAPINFO.SET_MOUSE_POSITION,
    lngLat,
    coordinate,
  };
}

/** Action to set the map extent.
 *  @param {number[]} extent Map extent in EPSG:4326.
 *
 *  @returns {Object} Action object to pass to reducer.
 */
export function setMapExtent(extent) {
  return {
    type: MAPINFO.SET_EXTENT,
    extent,
  };
}

/** Action to set the resolution.
 *  @param {number} resolution Resolution.
 *
 *  @returns {Object} Action object to pass to reducer.
 */
export function setResolution(resolution) {
  return {
    type: MAPINFO.SET_RESOLUTION,
    resolution,
  };
}

/** Action to set the projection.
 *  @param {string} projection Projection.
 *
 *  @returns {Object} Action object to pass to reducer.
 */
export function setProjection(projection) {
  return {
    type: MAPINFO.SET_PROJECTION,
    projection,
  };
}
