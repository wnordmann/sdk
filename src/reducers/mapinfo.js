/*
 * Copyright 2015-present Boundless Spatial Inc., http://boundlessgeo.com
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and limitations
 * under the License.
 */

/** @module reducers/mapinfo
 * @desc Map Info Reducer
 */

import {MAPINFO} from '../action-types';

const defaultState = {
  size: null,
  mouseposition: {
    lngLat: null,
    coordinate: null,
  },
};

/** Map info reducer.
 *  @param {Object} state The redux state.
 *  @param {Object} action The selected action object.
 *
 *  @returns {Object} The new state object.
 */
export default function mapInfoReducer(state = defaultState, action) {
  switch (action.type) {
    case MAPINFO.SET_SIZE:
      return Object.assign({}, state, {size: action.size});
    case MAPINFO.SET_MOUSE_POSITION:
      return Object.assign({}, state, {mouseposition: {lngLat: action.lngLat, coordinate: action.coordinate}});
    default:
      return state;
  }
}
