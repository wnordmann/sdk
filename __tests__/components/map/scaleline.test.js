/* global it, describe, expect, beforeEach */

import React from 'react';
import {mount, configure} from 'enzyme';
import  Adapter from 'enzyme-adapter-react-16';

import {createStore, combineReducers} from 'redux';

import MapReducer from '../../../src/reducers/map';
import MapInfoReducer from '../../../src/reducers/mapinfo';
import * as MapInfoActions from '../../../src/actions/mapinfo';
import * as MapActions from '../../../src/actions/map';
import ConnectedScaleLine, {ScaleLine} from '../../../src/components/map/scaleline';

configure({adapter: new Adapter()});

describe('Scaleline tests', () => {

  let store;

  beforeEach(() => {
    store = createStore(combineReducers({
      map: MapReducer,
      mapinfo: MapInfoReducer,
    }));
    store.dispatch(MapActions.setView([-98, 40], 4));
  });

  it('create the correct scaleline', () => {
    store.dispatch(MapInfoActions.setResolution(1000));
    const wrapper = mount(<ConnectedScaleLine store={store} />);
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('creates the meter suffix', () => {
    store.dispatch(MapInfoActions.setResolution(2.388657133911758));
    const wrapper = mount(<ConnectedScaleLine store={store} />);
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('creates the millimeter suffix', () => {
    store.dispatch(MapInfoActions.setResolution(0.009330691929342804));
    const wrapper = mount(<ConnectedScaleLine store={store} />);
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('creates the micrometer suffix', () => {
    store.dispatch(MapInfoActions.setResolution(0.000009330691929342804));
    const wrapper = mount(<ConnectedScaleLine store={store} />);
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('create the correct scaleline in EPSG:4326', () => {
    store.dispatch(MapInfoActions.setProjection('EPSG:4326'));
    store.dispatch(MapInfoActions.setResolution(1000));
    const wrapper = mount(<ConnectedScaleLine units='degrees' store={store} />);
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('create the correct scaleline in degrees', () => {
    store.dispatch(MapInfoActions.setResolution(19567.87924100512));
    const wrapper = mount(<ConnectedScaleLine units='degrees' store={store} />);
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('create the correct scaleline in degrees (minutes)', () => {
    store.dispatch(MapInfoActions.setResolution(300));
    const wrapper = mount(<ConnectedScaleLine units='degrees' store={store} />);
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('create the correct scaleline in degrees (seconds)', () => {
    store.dispatch(MapInfoActions.setResolution(1));
    const wrapper = mount(<ConnectedScaleLine units='degrees' store={store} />);
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('create the correct scaleline in imperial', () => {
    store.dispatch(MapInfoActions.setResolution(1000));
    const wrapper = mount(<ConnectedScaleLine units='imperial' store={store} />);
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('create the correct scaleline in imperial (ft)', () => {
    store.dispatch(MapInfoActions.setResolution(1));
    const wrapper = mount(<ConnectedScaleLine units='imperial' store={store} />);
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('create the correct scaleline in imperial (in)', () => {
    store.dispatch(MapInfoActions.setResolution(0.001));
    const wrapper = mount(<ConnectedScaleLine units='imperial' store={store} />);
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('create the correct scaleline in nautical', () => {
    store.dispatch(MapInfoActions.setResolution(1000));
    const wrapper = mount(<ConnectedScaleLine units='nautical' store={store} />);
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('create the correct scaleline in us', () => {
    store.dispatch(MapInfoActions.setResolution(1000));
    const wrapper = mount(<ConnectedScaleLine units='us' store={store} />);
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('create the correct scaleline in us (ft)', () => {
    store.dispatch(MapInfoActions.setResolution(1));
    const wrapper = mount(<ConnectedScaleLine units='us' store={store} />);
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('create the correct scaleline in us (in)', () => {
    store.dispatch(MapInfoActions.setResolution(0.001));
    const wrapper = mount(<ConnectedScaleLine units='us' store={store} />);
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('shouldComponentUpdate should work correctly', () => {
    store.dispatch(MapInfoActions.setResolution(1000));
    const props = {
      map: {
        center: [-98, 40],
        zoom: 4,
      },
      mapinfo: {
        projection: 'EPSG:3857',
        resolution: 1000,
      }
    };
    const wrapper = mount(<ScaleLine {...props} />);
    const scaleline = wrapper.instance();
    let result;
    result = scaleline.shouldComponentUpdate({mapinfo: {resolution: 1000}});
    expect(result).toBe(false);
    result = scaleline.shouldComponentUpdate({units: 'degrees'});
    expect(result).toBe(true);
    result = scaleline.shouldComponentUpdate({mapinfo: {resolution: 500}});
    expect(result).toBe(true);
    result = scaleline.shouldComponentUpdate({map: {center: [-98, 50]}});
    expect(result).toBe(true);
  });

});
