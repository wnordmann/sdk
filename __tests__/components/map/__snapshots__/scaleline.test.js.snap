// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Scaleline tests create the correct scaleline 1`] = `"<div class=\\"sdk-scale-line\\"><div class=\\"sdk-scale-line-inner\\" style=\\"width: 65px;\\">50 km</div></div>"`;

exports[`Scaleline tests create the correct scaleline in EPSG:4326 1`] = `"<div class=\\"sdk-scale-line\\"><div class=\\"sdk-scale-line-inner\\" style=\\"width: 100px;\\">100000 °</div></div>"`;

exports[`Scaleline tests create the correct scaleline in degrees (minutes) 1`] = `"<div class=\\"sdk-scale-line\\"><div class=\\"sdk-scale-line-inner\\" style=\\"width: 81px;\\">10 ′</div></div>"`;

exports[`Scaleline tests create the correct scaleline in degrees (seconds) 1`] = `"<div class=\\"sdk-scale-line\\"><div class=\\"sdk-scale-line-inner\\" style=\\"width: 81px;\\">2 ″</div></div>"`;

exports[`Scaleline tests create the correct scaleline in degrees 1`] = `"<div class=\\"sdk-scale-line\\"><div class=\\"sdk-scale-line-inner\\" style=\\"width: 74px;\\">10 °</div></div>"`;

exports[`Scaleline tests create the correct scaleline in imperial (ft) 1`] = `"<div class=\\"sdk-scale-line\\"><div class=\\"sdk-scale-line-inner\\" style=\\"width: 80px;\\">200 ft</div></div>"`;

exports[`Scaleline tests create the correct scaleline in imperial (in) 1`] = `"<div class=\\"sdk-scale-line\\"><div class=\\"sdk-scale-line-inner\\" style=\\"width: 66px;\\">2 in</div></div>"`;

exports[`Scaleline tests create the correct scaleline in imperial 1`] = `"<div class=\\"sdk-scale-line\\"><div class=\\"sdk-scale-line-inner\\" style=\\"width: 105px;\\">50 mi</div></div>"`;

exports[`Scaleline tests create the correct scaleline in nautical 1`] = `"<div class=\\"sdk-scale-line\\"><div class=\\"sdk-scale-line-inner\\" style=\\"width: 121px;\\">50 nm</div></div>"`;

exports[`Scaleline tests create the correct scaleline in us (ft) 1`] = `"<div class=\\"sdk-scale-line\\"><div class=\\"sdk-scale-line-inner\\" style=\\"width: 80px;\\">200 ft</div></div>"`;

exports[`Scaleline tests create the correct scaleline in us (in) 1`] = `"<div class=\\"sdk-scale-line\\"><div class=\\"sdk-scale-line-inner\\" style=\\"width: 66px;\\">2 in</div></div>"`;

exports[`Scaleline tests create the correct scaleline in us 1`] = `"<div class=\\"sdk-scale-line\\"><div class=\\"sdk-scale-line-inner\\" style=\\"width: 105px;\\">50 mi</div></div>"`;

exports[`Scaleline tests creates the meter suffix 1`] = `"<div class=\\"sdk-scale-line\\"><div class=\\"sdk-scale-line-inner\\" style=\\"width: 109px;\\">200 m</div></div>"`;

exports[`Scaleline tests creates the micrometer suffix 1`] = `"<div class=\\"sdk-scale-line\\"><div class=\\"sdk-scale-line-inner\\" style=\\"width: 70px;\\">500 μm</div></div>"`;

exports[`Scaleline tests creates the millimeter suffix 1`] = `"<div class=\\"sdk-scale-line\\"><div class=\\"sdk-scale-line-inner\\" style=\\"width: 70px;\\">500 mm</div></div>"`;

exports[`Scaleline tests should allow for custom className 1`] = `"<div class=\\"sdk-scale-line foo\\"><div class=\\"sdk-scale-line-inner\\" style=\\"width: 65px;\\">50 km</div></div>"`;
